File: python/vapor_idx/indexes/prefix.py
# =============================================================================
# vapor-idx — indexes/prefix.py
# Trie-based prefix index.
# Supports: startsWith
# =============================================================================

from __future__ import annotations
from dataclasses import dataclass, field
from typing import Any


@dataclass
class _TrieNode:
    ids:      set[str]          = field(default_factory=set)
    children: dict[str, "_TrieNode"] = field(default_factory=dict)


class PrefixIndex:
    def __init__(self) -> None:
        # field → trie root
        self._roots: dict[str, _TrieNode] = {}

    # ── Mutation ───────────────────────────────────────────────────────────────

    def add(self, field: str, value: Any, record_id: str) -> None:
        if value is None:
            return
        raw_values = value if isinstance(value, list) else [value]
        for raw in raw_values:
            normalised = str(raw).lower()
            if not normalised:
                continue
            if field not in self._roots:
                self._roots[field] = _TrieNode()
            node = self._roots[field]
            for char in normalised:
                if char not in node.children:
                    node.children[char] = _TrieNode()
                node = node.children[char]
                node.ids.add(record_id)

    def remove(self, field: str, value: Any, record_id: str) -> None:
        if value is None:
            return
        raw_values = value if isinstance(value, list) else [value]
        root = self._roots.get(field)
        if not root:
            return
        for raw in raw_values:
            normalised = str(raw).lower()
            node = root
            for char in normalised:
                child = node.children.get(char)
                if not child:
                    break
                child.ids.discard(record_id)
                node = child

    # ── Lookup ─────────────────────────────────────────────────────────────────

    def starts_with(self, field: str, prefix: str) -> set[str]:
        normalised = prefix.lower()
        root = self._roots.get(field)
        if not root:
            return set()
        node = root
        for char in normalised:
            child = node.children.get(char)
            if not child:
                return set()
            node = child
        return set(node.ids)

    # ── Lifecycle ──────────────────────────────────────────────────────────────

    def clear(self) -> None:
        self._roots.clear()

    @property
    def node_count(self) -> int:
        return sum(_count_nodes(root) for root in self._roots.values())


def _count_nodes(node: _TrieNode) -> int:
    return 1 + sum(_count_nodes(c) for c in node.children.values())

File: rust/prefix.rs
// =============================================================================
// vapor-idx — indexes/prefix.rs
// Trie-based prefix index.
// Supports: starts_with
// =============================================================================

use std::collections::{HashMap, HashSet};

#[derive(Debug, Default)]
struct TrieNode {
    ids:      HashSet<String>,
    children: HashMap<char, TrieNode>,
}

#[derive(Debug, Default)]
pub struct PrefixIndex {
    roots: HashMap<String, TrieNode>,
}

impl PrefixIndex {
    pub fn new() -> Self { Self::default() }

    pub fn add(&mut self, field: &str, value: &serde_json::Value, id: &str) {
        let texts: Vec<String> = match value {
            serde_json::Value::String(s) => vec![s.clone()],
            serde_json::Value::Array(arr) => arr.iter()
                .filter_map(|v| v.as_str().map(String::from))
                .collect(),
            _ => return,
        };

        let root = self.roots.entry(field.to_owned()).or_default();
        for text in texts {
            let normalised = text.to_lowercase();
            let mut node = root as *mut TrieNode;
            for ch in normalised.chars() {
                // SAFETY: single-threaded, no aliasing
                let n = unsafe { &mut *node };
                node = n.children.entry(ch).or_default() as *mut TrieNode;
                unsafe { (*node).ids.insert(id.to_owned()) };
            }
        }
    }

    pub fn remove(&mut self, field: &str, value: &serde_json::Value, id: &str) {
        let texts: Vec<String> = match value {
            serde_json::Value::String(s) => vec![s.clone()],
            serde_json::Value::Array(arr) => arr.iter()
                .filter_map(|v| v.as_str().map(String::from))
                .collect(),
            _ => return,
        };
        let Some(root) = self.roots.get_mut(field) else { return };
        for text in texts {
            let normalised = text.to_lowercase();
            let mut node = root as *mut TrieNode;
            for ch in normalised.chars() {
                let n = unsafe { &mut *node };
                let Some(child) = n.children.get_mut(&ch) else { break };
                child.ids.remove(id);
                node = child as *mut TrieNode;
            }
        }
    }

    pub fn starts_with(&self, field: &str, prefix: &str) -> HashSet<String> {
        let normalised = prefix.to_lowercase();
        let Some(root) = self.roots.get(field) else { return HashSet::new() };
        let mut node = root;
        for ch in normalised.chars() {
            let Some(child) = node.children.get(&ch) else { return HashSet::new() };
            node = child;
        }
        node.ids.clone()
    }

    pub fn clear(&mut self) { self.roots.clear(); }

    pub fn node_count(&self) -> usize {
        self.roots.values().map(count_nodes).sum()
    }
}

fn count_nodes(node: &TrieNode) -> usize {
    1 + node.children.values().map(count_nodes).sum::<usize>()
}
